onnxruntime/core/providers/webnn/builders/impl/slice_op_builder.cc

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Intel Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/safeint.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/optimizer/initializer.h"
+#include "core/providers/common.h"
+#include "core/providers/cpu/tensor/slice_helper.h"
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/webnn/builders/helper.h"
+#include "core/providers/webnn/builders/model_builder.h"
+#include "core/providers/webnn/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace webnn {
+
+class SliceOpBuilder : public BaseOpBuilder {
+  // Add operator related.
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const logging::Logger& logger) const override;
+  // TODO: Support Slice opset < 10, which uses attributes for starts and ends.
+  int GetMinSupportedOpSet(const Node& /* node */) const override { return 10; }
+};
+
+// Add operator related.
+
+void SliceOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  // Skip all initializer except the first inputs(data).
+  for (size_t i = 1; i < node.InputDefs().size(); i++) {
+    model_builder.AddInitializerToSkip(node.InputDefs()[i]->Name());
+  }
+}
+
+Status SliceOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                             const Node& node,
+                                             const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape;
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_shape, logger), "Cannot get shape");
+  auto rank = input_shape.size();
+  NodeAttrHelper helper(node);
+
+  emscripten::val inputs = model_builder.GetOperand(input_defs[0]->Name());
+  std::vector<int32_t> starts(rank, 0);
+  std::vector<int32_t> sizes(input_shape.begin(), input_shape.end());
```

Comment:
Doesn't the build complain about integer 64->32 truncation? The other places I've seen use transform to downcast.

Reply:
I didn't see the complain in build.

Reply:
Weird, maybe there are different build rules for the native code. 🤷‍♂️

Well I've seen this pattern frequently enough...

```
    std::transform(begin, end, output.begin, [](int64_t i) { return SafeInt<uint32_t>(i); });
```

...that it's probably worth adding a helper to downcast it. e.g. https://github.com/microsoft/onnxruntime/blob/5abaca9d6969f65998708abf6cbc7616036b1769/onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp#L245-L254